src/protocol/payload/tx.rs

```diff
@@ -0,0 +1,868 @@
+        self.prev_out_hash.encode(buffer)?;
+        buffer.write_all(&self.prev_out_index.to_le_bytes())?;
+
+        self.script_len.encode(buffer)?;
+        buffer.write_all(&self.script)?;
+
+        buffer.write_all(&self.sequence.to_le_bytes())?;
+
+        Ok(())
+    }
+
+    fn decode(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        let prev_out_hash = Hash::decode(bytes)?;
+        let prev_out_index = u32::from_le_bytes(read_n_bytes(bytes)?);
+
+        let script_len = VarInt::decode(bytes)?;
+        let mut script = vec![0u8; script_len.0];
+        bytes.read_exact(&mut script)?;
+
+        let sequence = u32::from_le_bytes(read_n_bytes(bytes)?);
+
+        Ok(Self {
+            prev_out_hash,
+            prev_out_index,
+            script_len,
+            script,
+            sequence,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct TxOut {
+    value: i64,
+    pk_script_len: VarInt,
+    pk_script: Vec<u8>,
+}
+
+impl TxOut {
+    fn encode(&self, buffer: &mut Vec<u8>) -> io::Result<()> {
+        buffer.write_all(&self.value.to_le_bytes())?;
+        self.pk_script_len.encode(buffer)?;
+        buffer.write_all(&self.pk_script)?;
+
+        Ok(())
+    }
+
+    fn decode(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        let value = i64::from_le_bytes(read_n_bytes(bytes)?);
+        let pk_script_len = VarInt::decode(bytes)?;
+        let mut pk_script = vec![0u8; pk_script_len.0];
+        bytes.read_exact(&mut pk_script)?;
+
+        Ok(Self {
+            value,
+            pk_script_len,
+            pk_script,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct JoinSplit {
+    pub_old: u64,
+    pub_new: u64,
+    anchor: [u8; 32],
+    // Two nullifiers are present, each 32 bytes long.
+    nullifiers: [u8; 64],
+    // Two commitments are present, each 32 bytes long.
+    commitments: [u8; 64],
+    ephemeral_key: [u8; 32],
+    random_seed: [u8; 32],
+    // Two tags are present, each 32 bytes long.
+    vmacs: [u8; 64],
+    // BCTV14 or Groth16, depending on the transaction version.
+    zkproof: Zkproof,
+    // Two cyphertex components are present, each 601 bytes long.
+    enc_cyphertexts: [u8; 1202],
+}
+
+impl JoinSplit {
+    fn encode(&self, buffer: &mut Vec<u8>) -> io::Result<()> {
+        buffer.write_all(&self.pub_old.to_le_bytes())?;
+        buffer.write_all(&self.pub_new.to_le_bytes())?;
+
+        buffer.write_all(&self.anchor)?;
+        buffer.write_all(&self.nullifiers)?;
+        buffer.write_all(&self.commitments)?;
+        buffer.write_all(&self.ephemeral_key)?;
+        buffer.write_all(&self.random_seed)?;
+        buffer.write_all(&self.vmacs)?;
+
+        self.zkproof.encode(buffer)?;
+        buffer.write_all(&self.enc_cyphertexts)?;
+
+        Ok(())
+    }
+
+    fn decode_bctv14(bytes: &mut Cursor<&[u8]>) -> io::Result<Self> {
+        // TODO: deduplicate (might require generics).
```

Comment:
Alternatives to generics:
1. add a boolean parameter to select which proof type to use. Not great if you wanted to have everything use the same `decode(bytes)` Trait eventually.
2. figure out which type it is by validating? I presume that's not really feasible or desirable though. 

Not really good ideas, sorry.

Reply:
> add a boolean parameter to select which proof type to use. Not great if you wanted to have everything use the same decode(bytes) Trait eventually.

Indeed, this is why I avoided it so far. Food for thought, not a pressing matter either way. I had also initially used the remaining length, though that's not really a guarentee we should be relying on.